libavcodec/rkmppdec.c

```diff
@@ -514,8 +514,8 @@ static int rkmpp_export_frame(AVCodecContext *avctx, AVFrame *frame, MppFrame mp
                     DRM_FORMAT_MOD_ARM_AFBC(AFBC_FORMAT_MOD_SPARSE | AFBC_FORMAT_MOD_BLOCK_SIZE_16x16);
         layer->format = rkmpp_get_drm_afbc_format(mpp_fmt);
         layer->nb_planes = 1;
-        layer->planes[0].pitch = mpp_frame_get_hor_stride(mpp_frame);
-        layer->planes[0].offset = mpp_frame_get_offset_y(mpp_frame) * layer->planes[0].pitch;
+        layer->planes[0].pitch = mpp_frame_get_hor_stride(mpp_frame) * 3 / 2;
```

Comment:
Did you get that value 1.5 from [here](https://github.com/JeffyCN/xorg-xserver/blob/0b0511a2ffe9406384ffd2f9892d4adfe9e74d47/glamor/glamor_xv.c#L672-L700)?

Reply:
I found that value by testing.
But i think it is more of a definition, mpp defines the stride as it is a semi-planar format with multiple planes, therefore bases the stride calculation in that regard and gives a MPP_FRAME_YUV420SP format with align(width)

But DRM expects a stride to represent the whole plane and since AFBC planes are single planes so Y + UV planes,  stride is 3/2 of the original stride for 420 subsampling, and *2 for 422 subsampling. (I have forgotten to distinguish against the subsampling here actually since it is a workaround).

On the other hand, i am not sure it is that simple for 4:2:0 case because, assume that you a width n*16, and your alignment is 16  (mpp's most cases)  and n is an odd number.

Ie: DVD resolution: 720x480
where n=45 (odd number): 45*16 = 720px.

for NV12 afbc your mpp stride is 720, but drm stride is 720*3/2 = 1080 which is no more 16 aligned but 8 aligned.

So as the consumer of this frames are drm, this also depends of the VOP2 alignment or whatever the consumer is requirement, but for encoder cases this is a risk, so therefore i consider this as an mpp bug.

Entry is here
https://github.com/rockchip-linux/mpp/issues/509

Tagging @FumasterLin @JeffyCN @HermanChen

Reply:
> but for encoder cases and some whatever future case this is a risk

I think I may have encountered somthing similar, but the width of my resolution is not 16-aligned, for example, 2900x2160 (slightly different from 2880x2160, 4:3). Neither the MPP encoder nor the RGA can handle it correctly.

The MPP HEVC decoder (AFBC) gives a stride 2944 (186x16 or 46x64)

2944*3/2=4416 (aligned with 8/16/64, but the broken image suggests a wrong stride)

<img src="https://github.com/nyanmisaka/ffmpeg-rockchip/assets/14953024/58d2213b-06b4-4a66-abe5-c0a71721cf8d" width="300">

Reply:
I think this is 64 aligned, 2944 does not align to 16, yet, i have created a test hevc nv12 file 2900x2160 ir decodes correctly and gives 2944*3/2 = 4416 to layer descriptor.

However i think mpp again needs to scale down this 2/3 or RGA as well, because they expect only UV plane strides regardless that the picture is AFBC or not. That i have not PRed yet if you are using this patchset.

Reply:
For AFBC input in RGA and MPP encoders, they don't care about the given stride, but use the given width and the hardcoded 16 ([h264e](https://github.com/rockchip-linux/mpp/blob/0e03d3bc704da949da044519c32ecf9b6e09de09/mpp/hal/rkenc/h264e/hal_h264e_vepu580.c#L721), [h265e](https://github.com/rockchip-linux/mpp/blob/0e03d3bc704da949da044519c32ecf9b6e09de09/mpp/hal/rkenc/h265e/hal_h265e_vepu580.c#L1944)) to calculate stride.

Try [this](https://github.com/nyanmisaka/ffmpeg-rockchip/commits/use-crop/) branch

```
./ffmpeg -f lavfi -i testsrc=s=2900x2160,format=yuv420p -c:v hevc_rkmpp -b:v 6M -vframes 100 -y /tmp/testsrc_2900x2160_hevc_8bit.mp4

./ffmpeg -hwaccel rkmpp -hwaccel_output_format drm_prime -afbc 1 -i /tmp/testsrc_2900x2160_hevc_8bit.mp4 -vf scale_rkrga=format=nv12:afbc=1 -c:v hevc_rkmpp -b:v 6M -y /tmp/broken.mp4
```

And you will get
<img src="https://github.com/nyanmisaka/ffmpeg-rockchip/assets/14953024/ae10fead-2360-4eb5-932f-e29cd203206c" width="500">

Reply:
> or AFBC input in RGA and MPP encoders, they don't care about the given stride, but use the given width and the hardcoded 16 ([h264e](https://github.com/rockchip-linux/mpp/blob/0e03d3bc704da949da044519c32ecf9b6e09de09/mpp/hal/rkenc/h264e/hal_h264e_vepu580.c#L721), [h265e](https://github.com/rockchip-linux/mpp/blob/0e03d3bc704da949da044519c32ecf9b6e09de09/mpp/hal/rkenc/h265e/hal_h265e_vepu580.c#L1944)) to calculate stride.

If not given, because, you can never know the stride is actually the one you estimate, you have to know this to actually process the image as is, because each pixels are written in the given stride size, if it does not align it will skew.

So for enc, setting the afbc_hor_stride of the mppframe should fix the issue

FFLAIGN(2900, 16) = 2912 != FFLAIGN(2900, 64) = 2944

Reply:
```patch
diff --git a/libavcodec/rkmppdec.c b/libavcodec/rkmppdec.c
index d155278..48cd619 100644
--- a/libavcodec/rkmppdec.c
+++ b/libavcodec/rkmppdec.c
@@ -503,7 +503,8 @@
         layer->format = rkmpp_get_drm_afbc_format(mpp_fmt);
         layer->nb_planes = 1;
         layer->planes[0].offset = 0;
-        layer->planes[0].pitch  = mpp_frame_get_hor_stride(mpp_frame);
+        // TODO: handle scale factor per subsampling
+        layer->planes[0].pitch  = mpp_frame_get_hor_stride(mpp_frame) * 3 / 2;
 
         /* MPP specific AFBC src_x|y offsets, not memory address offsets */
         frame->crop_top  = mpp_frame_get_offset_y(mpp_frame);
diff --git a/libavcodec/rkmppenc.c b/libavcodec/rkmppenc.c
index d9ad512..843a2f1 100644
--- a/libavcodec/rkmppenc.c
+++ b/libavcodec/rkmppenc.c
@@ -88,7 +88,7 @@
     plane1 = &layer->planes[1];
 
     *hs = plane0->pitch;
-    *vs = is_packed_fmt ?
+    *vs = (is_packed_fmt || layer->nb_planes == 1) ?
         ALIGN_DOWN(object->size / plane0->pitch, is_rgb ? 1 : 2) :
         (plane1->offset / plane0->pitch);
 
@@ -523,8 +523,7 @@
 
     is_afbc = drm_is_afbc(drm_desc->objects[0].format_modifier);
     if ((r->pix_fmt == AV_PIX_FMT_YUV420P ||
-         r->pix_fmt == AV_PIX_FMT_YUV422P ||
-         is_afbc) && (drm_frame->width % 2)) {
+         r->pix_fmt == AV_PIX_FMT_YUV422P) && (drm_frame->width % 2)) {
         av_log(avctx, AV_LOG_ERROR, "Unsupported width %d, not 2-aligned\n", drm_frame->width);
         goto exit;
     }
@@ -539,19 +538,23 @@
     mpp_frame_set_color_range(mpp_frame, avctx->color_range);
 
     pix_desc = av_pix_fmt_desc_get(r->pix_fmt);
-    if (!is_afbc) {
-        ret = get_byte_stride(&drm_desc->objects[0],
-                              &drm_desc->layers[0],
-                              (pix_desc->flags & AV_PIX_FMT_FLAG_RGB),
-                              (pix_desc->flags & AV_PIX_FMT_FLAG_PLANAR),
-                              &hor_stride, &ver_stride);
-        if (ret < 0 || !hor_stride || !ver_stride) {
-            av_log(avctx, AV_LOG_ERROR, "Failed to get frame strides\n");
-            goto exit;
-        }
+    ret = get_byte_stride(&drm_desc->objects[0],
+                          &drm_desc->layers[0],
+                          (pix_desc->flags & AV_PIX_FMT_FLAG_RGB),
+                          (pix_desc->flags & AV_PIX_FMT_FLAG_PLANAR),
+                          &hor_stride, &ver_stride);
+    if (ret < 0 || !hor_stride || !ver_stride) {
+        av_log(avctx, AV_LOG_ERROR, "Failed to get frame strides\n");
+        goto exit;
+    }
 
-        mpp_frame_set_hor_stride(mpp_frame, hor_stride);
-        mpp_frame_set_ver_stride(mpp_frame, ver_stride);
+    mpp_frame_set_hor_stride(mpp_frame, hor_stride);
+    mpp_frame_set_ver_stride(mpp_frame, ver_stride);
+    if(is_afbc){
+        // TODO: handle scale factor per sub-sampling basis
+        if(pix_desc->flags & AV_PIX_FMT_FLAG_PLANAR)
+            hor_stride = hor_stride * 2 / 3;
+        mpp_frame_set_fbc_hdr_stride(mpp_frame, hor_stride);
     }
 
     buf_info.type  = MPP_BUFFER_TYPE_DRM;
```

so above works, could be better implemented may be.

The illusion here is that 64 byte align stride is only 16byte aligned when the actual width is a multiple of 16.

Reply:
>The only point I didn't get is the difference between FBC header stride and horizontal stride.

I think mpp guys tried to mean HDR (like 10bit NV15 & NV20) not header, and fbc and 10bit or even sometimes if width > 4k or something, mpp spits 256 odd align or 64 align or whatever it pleases. Normally it is 16. So i think using `mpp_frame_get_fbc_hdr_stride` in afbc case, in other cases `mpp_frame_get_hor_stride` makes sense to me.

Also side note: I dont think below is reliable in all cases in rkmppenc.

```c
plane1->offset / plane0->pitch
```

because there is no guarantee that the offset is going to be contiguous from the previous Y plane, even if it points to the same drm object.

ie yuv420p: actual organization in buffer Y + V + U
Plane0=start0: Y Plane
Plane1=Start(sizeof(V+Y): U plane)
Plane2=Start(sizeof(Y): V plane)

or in case each plane refer to different objects then their offsets will be 0.

Reply:
Isn't `AV_PIX_FMT_YUV420P` actually `YU12` (Y + U/Cb + V/Cr)?

(Y + V/Cr + U/Cb) is `YV12`, which is [not directly supported in FFmpeg](https://stackoverflow.com/questions/46930307/does-ffmpeg-support-yv12) but you can find it in drm fourcc.

```
/*
 * 3 plane YCbCr
 * index 0: Y plane, [7:0] Y
 * index 1: Cb plane, [7:0] Cb
 * index 2: Cr plane, [7:0] Cr
 * or
 * index 1: Cr plane, [7:0] Cr
 * index 2: Cb plane, [7:0] Cb
 */
#define DRM_FORMAT_YUV420	fourcc_code('Y', 'U', '1', '2') /* 2x2 subsampled Cb (1) and Cr (2) planes */
#define DRM_FORMAT_YVU420	fourcc_code('Y', 'V', '1', '2') /* 2x2 subsampled Cr (1) and Cb (2) planes */
```

https://github.com/torvalds/linux/blob/052d534373b7ed33712a63d5e17b2b6cdbce84fd/include/uapi/drm/drm_fourcc.h#L393

Reply:
I mean it was just an example, because there is no guarantee that the buffer will have 1 single plane, Y+U+V contiguously placed on top of each other. But you are right as long as picture is single buffer it is very odd to have  unnecessary data in between planes, or mixed up order is not possible. But having multiple objects per plane is very likely. There is FF_CAP_CONTIGOUS_BUFFERS flag in FFmpeg to ensure this, but only old rkmppdec uses it, none of the other codecs take care of it.

So it depends on the allocator behaviour. And in case V4l2 or other drm prime based codecs this might happen.

Reply:
May be an extra check like nb_objects == 1 would be enough, rest is funny confuguirations no need to support.

Reply:
DRM Prime Buffer imported from external HDMI RX (v4l2) or kmsgrab may indeed be non-contigous.

Reply:
rkmppenc.c
```c
    *vs = (is_packed_fmt || layer->nb_planes == 1) ?
         ALIGN_DOWN(object->size / plane0->pitch, is_rgb ? 1 : 2) :
         (plane1->offset / plane0->pitch);
```

oh boy, also `object->size/plane->pitch` is also not reliable, there is no guarantee that the objects buffers is allocated in the size of the image, it can also a be a bigger buffer which was reused. In this case hstride will be wrong as well.

I think hsride is easy, it will be layer0->pitch in most cases, but for vstride, i am starting to think to directly use height without checking any alignment in the decoder.

Reply:
> oh boy, also object->size/plane->pitch is also not reliable, there is no guarantee that the objects buffers is allocated in the size of the image, it can also a be a bigger buffer which was reused. In this case vstride will be wrong as well.

> I think hsride is easy, it will be layer0->pitch in most cases, but for vstride, i am starting to think to directly use height without checking any alignment in the encoder.

> Using directly height will work for 1 plane formats as long as hstride xheight x bpp <= the buffer size.

<img src="https://github.com/nyanmisaka/ffmpeg-rockchip/assets/14953024/4f234476-0f35-487c-9070-5630956eedd4" width="500">

For non-packed/single plane image, IMHO it only needs to satisfy:
`ver_stride >= height` and `hor_stride * ver_stride * bpp <= buffer_size`.

There is no doubt that using `height` as `ver_stride` in this case is more accurate, but from my previous testing, MPPenc and RGA are not picky about it. One exception is that RGA has some special requirements, [packed YUYV422](https://github.com/airockchip/librga/blob/main/docs/Rockchip_Developer_Guide_RGA_EN.md#image-format-alignment-instructions) requires `ver_stride` to be 2-aligned or it fails at checking stage.

> RK_FORMAT_YUYV_422
RK_FORMAT_YVYU_422
RK_FORMAT_UYVY_422
RK_FORMAT_VYUY_422
width stride must be 8-aligned，x_offset、y_offset、width、height、**height stride** must be 2-aligned

---

> So if there is a case where the buffer is single but there are multiple planes, and the 2nd plane (or 3rd) offset is not starting from hstride x vstride x bbp, or in otherwords there are some missing buffers sizes in between planes, normally this is because of the alignment, so offset1/hstride0 must give the vstride. (and the missing fill buffers are alignment paddings)

> If not there is something else stored in the buffer, then we can not feed this to mpp, because it only accepts contigous planes. So this plane must be copied to a new contigous buffer. And i was doing this with libyuv, (now i remembered)...

This is mentioned in the [MPP docs](https://github.com/rockchip-linux/mpp/blob/develop/doc/Rockchip_Developer_Guide_MPP_EN.md#361-widthand-height-of-input-image-and-stride), and users need to fill in the padding areas or copy the planes to a new contigous buffer.

<img src="https://github.com/nyanmisaka/ffmpeg-rockchip/assets/14953024/e8c54cf1-2281-4aac-be9a-d84787e4989b" width="250"><img src="https://github.com/nyanmisaka/ffmpeg-rockchip/assets/14953024/6304591b-ee1f-45a9-a568-59ae375b4a96" width="250">

Here are the following situations:
- drm_prime frames produced by MPPdec: assuming they meet the requirements of RGA and MPPenc (alignments and contigous), although MPPdec still lacks the functionality to set `hor_stride`.

- Software frames: perform copy to hwcontext_rkmpp frame_pool to meet the needs of RGA and MPPenc (such as `testsrc, hwupload` and rkmppenc.c [enforces copying](https://github.com/nyanmisaka/ffmpeg-rockchip/blob/2938ac0758883a31e41e02590dd913ce5b79cb60/libavcodec/rkmppenc.c#L503) for software frames)

- drm_prime frames imported from external API: If the requirements of RGA and MPPenc are not met, the behavior is undefined. Currently, neither RGA nor MPPenc can handle this situation.

Reply:
@nyanmisaka 
https://github.com/hbiyik/ffmpeg-rockchip/commit/94e3d02752b6ae1551dc3f687e178e0b2e4d2539

I did another test, and i need to rectify, mpp's original afbc stride is based on hor_stride, not afbc_hdr_stride. With you afbc refactoring 8k files give wrong picture, it is fixed by applying the stride calculation based on hor_stride, but now rga calculation is messed. I am actually quite confused again but sure, that decode must base 3/2, 2 or 3 multiplier to hor_stride not afbc_hdr stride.

Reply:
@hbiyik 
IMHO we just need to add the following lines or reverse to RGA input/output and MPPenc input.

```c
        if (pix_fmt == AV_PIX_FMT_NV15 || pix_fmt == AV_PIX_FMT_NV20)
            layer->planes[0].pitch = layer->planes[0].pitch * 8 / 10;
```

Reply:
ok i think i did it this time. bytes_pp already includes the plane multipler x bit depth multiplier in it.  So below seems to fix rga as well.

https://github.com/hbiyik/ffmpeg-rockchip/commit/1b21b96aa0a214ca6e0e4f7f2d3976e8b5e3e6f8

Tested with decoder(afbc) with kodi
Tested with decoder(afbc)->rga(plain nv12/16) formats with mpv.

So it should work the other way around as well.

Reply:
> bytes_pp already includes the plane multipler x bit depth multiplier in it.

It does not. It happens to work only because [these lines of code](https://github.com/hbiyik/ffmpeg-rockchip/commit/1b21b96aa0a214ca6e0e4f7f2d3976e8b5e3e6f8#diff-8fac2cc4c170d4d20db09b4a28f0ff1bd1046b33142e600d3ef97938ad0ce1c6R473) roll back `wstride` to use `ALIGN(width, 16)`.

Can you test these in mpv with [this branch](https://github.com/nyanmisaka/ffmpeg-rockchip/commits/use-crop/)?
decoder(afbc) nv12/nv16/nv15/nv20 -> rga(afbc=0) nv12/nv16/p010/p210

We lack a player capable of testing RGA(afbc) output, otherwise we can test these:
decoder(afbc) nv12/nv16/nv15/nv20 -> rga(afbc=1) nv12/nv16/nv15/nv20/bgra/bgr565

Reply:
it seems that it really does but i was too sleepy to [multiple instead of dividing](https://github.com/hbiyik/ffmpeg-rockchip/commit/7af1bc3ff3137f1aa74d354ef49bee24e0c3d1f3). and commented out [these lines of code](https://github.com/hbiyik/ffmpeg-rockchip/commit/1b21b96aa0a214ca6e0e4f7f2d3976e8b5e3e6f8#diff-8fac2cc4c170d4d20db09b4a28f0ff1bd1046b33142e600d3ef97938ad0ce1c6R473) to make sure as well.

here are the gdb outputs of bytes_pp

![image](https://github.com/nyanmisaka/ffmpeg-rockchip/assets/24766436/4858c6f0-9a43-43b7-9b33-000cdbcc3a8e)
![image](https://github.com/nyanmisaka/ffmpeg-rockchip/assets/24766436/c3385157-99b8-40b9-bded-58940b1bcb49)
![image](https://github.com/nyanmisaka/ffmpeg-rockchip/assets/24766436/bdfcaa02-a9cd-4475-9ede-4883c5dda81d)

Reply:
> We lack a player capable of testing RGA(afbc) output, otherwise we can test these:
> decoder(afbc) nv12/nv16/nv15/nv20 -> rga(afbc=1) nv12/nv16/nv15/nv20/bgra/bgr565

I might create a python script interfacing with KMS to display only 1 frame sometime but not soon. It is really a pain to work with mpv and kodi and vop2 when they all have some bits missing, there is no proper reference.

Reply:
Well, I only tested NV20 and it gave me a wrong value 4.

We may need to update the AVPixFmtDescriptor for the so called nv20le/be.
```c
    [AV_PIX_FMT_NV20LE] = {
        .name = "nv20le",
        .nb_components = 3,
        .log2_chroma_w = 1,
        .log2_chroma_h = 0,
        .comp = {
            { 0, 2, 0, 0, 10 },        /* Y */
            { 1, 4, 0, 0, 10 },        /* U */
            { 1, 4, 2, 0, 10 },        /* V */
        },
        .flags = AV_PIX_FMT_FLAG_PLANAR,
    },

    [AV_PIX_FMT_NV15] = {
        .name = "nv15",
        .nb_components = 3,
        .log2_chroma_w = 1,
        .log2_chroma_h = 1,
        .comp = {
            { 0, 10,  0, 0, 10 },      /* Y */
            { 1, 20,  0, 0, 10 },      /* U */
            { 1, 20, 10, 0, 10 },      /* V */
        },
        .flags = AV_PIX_FMT_FLAG_PLANAR | AV_PIX_FMT_FLAG_BITSTREAM,
    },
```

Reply:
Ah i remember that we had discussed this you remeber?, it was wrong in ffmpeg already. it should be defined with `AV_PIX_FMT_FLAG_BITSTREAM` this actual definition is something like P210 actually that.

Reply:
Even having LE or BE is absurd, you can not have an LE or BE if you do not have padding, unless they are talking about another NV20 which is different than Rockchip's NV20. I hope not.

Reply:
Sure. Changing like this fixes the value. It gives me 2.5 (2.0*1.25) now.
```c
    [AV_PIX_FMT_NV20LE] = {
        .name = "nv20le",
        .nb_components = 3,
        .log2_chroma_w = 1,
        .log2_chroma_h = 0,
        .comp = {
            { 0, 10,  0, 0, 10 },      /* Y */
            { 1, 20,  0, 0, 10 },      /* U */
            { 1, 20, 10, 0, 10 },      /* V */
        },
        .flags = AV_PIX_FMT_FLAG_PLANAR | AV_PIX_FMT_FLAG_BITSTREAM,
    },
```

Does it look correct to you? Also, is the nv20BE variant necessary?

Reply:
> Even having LE or BE is absurd, you can not have an LE or BE if you do not have padding, unless they are talking about another NV20 which is different than Rockchip's NV20. I hope not.

Feel the same. If nv20be is not supposed to exist, then we should mark it as deprecated.

Reply:
[AV_PIX_FMT_NV20LE -> AV_PIX_FMT_NV20] = {
        .name = "nv20le",
        .nb_components = 3,
        .log2_chroma_w = 1,
        .log2_chroma_h = 0,
        .comp = {
            { 0, 10,  0, 0, 10 },      /* Y */
            { 1, 20,  0, 0, 10 },      /* U */
            { 1, 20, 10, 0, 10 },      /* V */
        },
        .flags = AV_PIX_FMT_FLAG_PLANAR | AV_PIX_FMT_FLAG_BITSTREAM,
    },

I think this definition is correct, i am also not the expert but seems rgith to me.
But i think it should be directly `AV_PIX_FMT_NV20`, and there should be no `AV_PIX_FMT_NV20LE` nor `AV_PIX_FMT_NV20BE`, because there is no such thing.

Here is the reference: https://github.com/JeffyCN/FFmpeg/issues/18#issuecomment-1824587867

A side topic: Do you think this hack here could be related to this wrong definition. https://github.com/nyanmisaka/ffmpeg-rockchip/commit/c359d00d131770502deaf1c198211edcf43fa013

Reply:
>  But i think it should be directly `AV_PIX_FMT_NV20`, and there should be no `AV_PIX_FMT_NV20LE` nor `AV_PIX_FMT_NV20BE`, because there is no such thing.

Now when i think about it, removing BE and LE variants might introduce API change and the array size will change, this can be dangerous, so yeah you are right, may be just point BE or LE to the same definition would work.

Reply:
Since the enum of NV20LE/BE was defined before NV20, we cannot point NV20LE/BE to NV20.
We should remove the nv20->nv20le/be shortcut and introduce a new AV_PIX_FMT_NV20 format.
```c
// pixfmt.h
#define AV_PIX_FMT_NV20       AV_PIX_FMT_NE(NV20BE,  NV20LE)
```

Reply:
> > bytes_pp already includes the plane multipler x bit depth multiplier in it.
> 
> It does not. It happens to work only because [these lines of code](https://github.com/hbiyik/ffmpeg-rockchip/commit/1b21b96aa0a214ca6e0e4f7f2d3976e8b5e3e6f8#diff-8fac2cc4c170d4d20db09b4a28f0ff1bd1046b33142e600d3ef97938ad0ce1c6R473) roll back `wstride` to use `ALIGN(width, 16)`.
> 
> Can you test these in mpv with [this branch](https://github.com/nyanmisaka/ffmpeg-rockchip/commits/use-crop/)? decoder(afbc) nv12/nv16/nv15/nv20 -> rga(afbc=0) nv12/nv16/p010/p210
> 
> We lack a player capable of testing RGA(afbc) output, otherwise we can test these: decoder(afbc) nv12/nv16/nv15/nv20 -> rga(afbc=1) nv12/nv16/nv15/nv20/bgra/bgr565

FYI， the afbc nv12/nv16/nv15/nv20  output from rga3 is not compatible with vop。

Reply:
> FYI， the afbc nv12/nv16/nv15/nv20 output from rga3 is not compatible with vop。

@andyshrk 
Thx for the info. This is optional for our use case.

But can you elaborate on this?

Also, is [this kernel commit](https://github.com/rockchip-linux/kernel/commit/a662526fcbe1281b7dcdb74c136feb87eab99646) related?

Reply:
> > FYI， the afbc nv12/nv16/nv15/nv20 output from rga3 is not compatible with vop。
> 
> @andyshrk Thx for the info. This is optional for our use case.
> 
> But can you elaborate on this?
> 
> Also, is [this kernel commit](https://github.com/rockchip-linux/kernel/commit/a662526fcbe1281b7dcdb74c136feb87eab99646) related?

Not， it has nothing to do with this commit，it‘s because of the rga IC design，but I don’t have the detail information。I just heard this from other people。

Reply:
> Not， it has nothing to do with this commit，it‘s because of the rga IC design，but I don’t have the detail information。I just heard this from other people。

Got it. Thx anyway.

Reply:
> > A side topic: Do you think this hack here could be related to this wrong definition.
> 
> This works now with the new NV20.
> 

Perfect, because as next step i was planning on implementing NEON acceleration on plane copy, and this hack was making things uglier.

Reply:
> Perfect, because as next step i was planning on implementing NEON acceleration on plane copy, and this hack was making things uglier.

Done in #17